### PR TITLE
print sprintf testing rule

### PR DESCRIPTION
**Pull request by Ronnie-personal**
Adding dubious-print-sprintf under testing category.

This PR is to address #220

### CODE REVIEW

**Comment by Ronnie-personal** [comment]
@anderseknert 
I guess I still need to work on readme file and data.yaml.

Does the code changes overall make sense to you?

Also I'm not sure why 'config_test.rego' failed, that's not part of this PR code change.

Thanks!

**Comment by anderseknert** [comment]
The `config_test.rego` test checks that all rules have an entry in the `data.yaml` file, which like you said you still need to to :) There's an [open issue](https://github.com/StyraInc/regal/issues/231) to have the `regal new rule` command add this automatically, along with a page for documentation of the rule. If you're up for it later, that'd be a good one to work on!

Other than that, the code looks good to me! The way it currently works however means it'll only find `print` statements in the top level of a rule — i.e. if you add a test for this rule, you'll see that it won't fail:

```rego
y {
    comprehension := [x |
        x := input[_]
        print(sprintf("x is: %s", [x]))
    ]
}
```

To fix this, you could use the `walk` built-in function to traverse the whole rule. See the code for [non-raw-regex-pattern](https://github.com/StyraInc/regal/blob/main/bundle/regal/rules/idiomatic/non_raw_regex_pattern.rego) for an example, but basically this:

```rego
report contains violation if {
	# regal ignore:unused-return-value,function-arg-return
	walk(input.rules, [_, value])

	value[0].type == "ref"
	value[0].value[0].type == "var"
	value[0].value[0].value == "print"

        # the rest of your code here
}
```

Traversing everything like this is quite slow, and I plan to investigate alternative approaches in the future. But let's go with that for now as at least it will be correct :)

For the docs, I think the description in the issue would work well: https://github.com/StyraInc/regal/issues/220

File: bundle/regal/rules/testing/dubious_print_sprintf.rego [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# METADATA
+# description: dubious print sprintf
```
Comment (anderseknert) [inline bundle/regal/rules/testing/dubious_print_sprintf.rego:2]: Let's change this to "Dubious use of `print` and `sprintf`"

Comment (Ronnie-personal) [reply bundle/regal/rules/testing/dubious_print_sprintf.rego:2]: Thanks for the comment. I will fix it.

File: bundle/regal/rules/testing/dubious_print_sprintf.rego [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# METADATA
+# description: dubious use of print and sprintf
```
Comment (anderseknert) [inline bundle/regal/rules/testing/dubious_print_sprintf.rego:2]: dubious -> Dubious

File: bundle/regal/rules/testing/dubious_print_sprintf_test.rego [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+			"ref": config.docs.resolve_url("$baseUrl/$category/dubious-print-sprintf", "testing"),
+		}],
+		"title": "dubious-print-sprintf",
+	}}
+}
+
+test_fail_bodies_print_sprintf if {
+	test_policy := ast.policy(`
+y {
+comprehension := [x |
```
Comment (anderseknert) [inline bundle/regal/rules/testing/dubious_print_sprintf_test.rego:37]: Please indent the body of this comprehension

File: e2e/testdata/violations/most_violations.rego [hunk 4]
```diff
@@ -114,3 +114,8 @@ use_some_for_output_vars {
 
 # metasyntactic variable
 foo := "bar"
+
+# dubious print sprintf
+y {
+print(sprintf("name is: %s domain is: %s", [input.name, input.domain]))
```
Comment (anderseknert) [inline e2e/testdata/violations/most_violations.rego:120]: Indent

**Review by anderseknert — COMMENTED** [review]
Looks good! Just some small fixes and then we can have this merged 👍

File: bundle/regal/rules/testing/dubious_print_sprintf_test.rego (hunk 3 above)
Comment (Ronnie-personal) [reply bundle/regal/rules/testing/dubious_print_sprintf_test.rego:37]: Sure, will do.

**Review by anderseknert — APPROVED** [review]
This is looking really good! There's still the docs to have added, but I can take care of that in a follow-up fix.

Thanks for contributin to Regal! 👏

**Pull request merged**